lib/module.js

```diff
@@ -2,10 +2,11 @@
 
 const { findSourceMap } = require('internal/source_map/source_map_cache');
 const { Module } = require('internal/modules/cjs/loader');
-const { register } = require('internal/modules/esm/loader');
+const { releaseLoadedModule, register } = require('internal/modules/esm/loader');
 const { SourceMap } = require('internal/source_map/source_map');
 
 Module.findSourceMap = findSourceMap;
 Module.register = register;
+Module.releaseLoadedModule = releaseLoadedModule;
```

Comment:
This cannot be available to the main thread as its effects are forbidden by ECMAScript spec. The best we can do is to have it available only in the loader thread.

Reply:
Can you explain? Is this just a way of saying that hooks need to be registered before users are allowed to use this? Why?

How would this work from the other thread since the caches are on the main thread? Cross thread communication?

Reply:
For reference, the relevant specification is https://tc39.es/ecma262/#sec-HostLoadImportedModule:
> - If this operation is called multiple times with the same (referrer, specifier) pair and it performs [FinishLoadingImportedModule](https://tc39.es/ecma262/#sec-FinishLoadingImportedModule)(referrer, specifier, payload, result) where result is a [normal completion](https://tc39.es/ecma262/#sec-completion-record-specification-type), then it must perform [FinishLoadingImportedModule](https://tc39.es/ecma262/#sec-FinishLoadingImportedModule)(referrer, specifier, payload, result) with the same result each time.

Reply:
Yes, we're aware that this API explicitly violates spec if used.

Reply:
It means that the spec gives some guarantees that `await import("a")` will always return the same value from the same module (as long as the module resolution has succeeded at least once)

Reply:
Yes, I know, that was the discussion that preceded this. There are use cases for violating this spec, like hot module reload.

Reply:
Attaching it to `Module`, though, would allow any code in your entire application to do so - surely it's best to restrict the functionality to loaders?

Reply:
@aduh95 thanks for raising your concern. I brought up this exact point, almost word for word, in this comment: https://github.com/nodejs/node/issues/49442#issuecomment-1793268584
> This breaks the invariants of `HostLoadImportedModule`: "If this operation is called multiple times with the same (referrer, specifier) pair [then it must resolve] with the same result each time". I suppose the difference here is that we are proposing a globally-importable utility which could be used outside of a loader.
> [...]
> The risks are impossible to enumerate since we're reneging on an presumed invariant. Like what happens if a user attempts to evict `node:fs`? Idk, will it segfault?

And also here https://github.com/nodejs/node/issues/49442#issuecomment-1740995593:
> Implementing this feature in a blessed way in nodejs is not difficult but doing so may be in [direct violation of the specification](https://tc39.es/ecma262/#sec-HostLoadImportedModule):
> 
> > If this operation is called multiple times with the same (referrer, specifier) pair and it performs [FinishLoadingImportedModule](https://tc39.es/ecma262/#sec-FinishLoadingImportedModule)(referrer, specifier, payload, result) where result is a [normal completion](https://tc39.es/ecma262/#sec-completion-record-specification-type), then it must perform [FinishLoadingImportedModule](https://tc39.es/ecma262/#sec-FinishLoadingImportedModule)(referrer, specifier, payload, result) with the same result each time.
> 
> That might be ok because nodejs has lots of power tools that alter the fabric of reality.
> 
> What I don't like about `delete require.cache[key]` and my sample above is that they provide the capability for _any_ module to remove _any other_ module. That was ok during the complete anarchy of CommonJS but ESM should probably hold itself to a higher standard. I haven't thought about this much at all, and this is an idea off the top of my head, but someone might consider implementing something like `import.meta.releaseSelf()` which would allow a module to release itself from the module graph. In the case of an error during dynamic import you could define the release function as a property on the error object itself [and all consumers of an errored dynamic module would need to release]. That would isolate the powerful capability to known sites. Or maybe it belongs as an [import attribute](https://github.com/tc39/proposal-import-attributes) i.e. `import('./maybe.mjs', { with: { weak: true } })`, and is only allowed on dynamic imports.

Here https://github.com/nodejs/node/issues/49442#issuecomment-1787708180:
> Yes, I would discourage its use. `require.register` and `require.cache` contributed to the total anarchy we've seen under CommonJS that the ecosystem is still feeling to this day. The evict function explicitly breaks the invariants promised under the specification. I'm clearly ok with reality-altering features (see [fibers](https://www.npmjs.com/package/fibers)) but I do want to make sure I communicate the nature of what I've proposed.

And here https://github.com/nodejs/node/issues/49442#issuecomment-1792763730:
> The language in the specification is clear, and our proposed function is a violation of the invariants expected of the host. As a matter of personal style I am totally ok violating any rule for any reason, but I do want to make sure we're on the same page that this is a dangerous and powerful violation of the specification.

I also suggested an alternate approach maybe with import attributes here: https://github.com/nodejs/node/issues/49442#issuecomment-1787593660

It is worth pointing that loaders already violate the specification because the result of `resolve` isn't internally memoized nor is it pure. Therefore the result of the specified `HostLoadImportedModule` is not necessarily the same given an identical [referrer, specifier] pair.

I'm happy to only enable this in loaders, but this cache lives in the main thread and therefore it is an operation which must take place in the main thread. Since the loaders API has moved to a `register` pattern which can be mutated at any time it is not totally clear how to keep this out of the hands of non-loader modules.

Reply:
It’s my opinion – and I believe the consensus in the project – that Node.js, as a JavaScript runtime, should parse JS source files according to the spec. Loaders can of course choose to parse them as something else, and deviate from the spec / follow a different one even. 
If Node.js starts to deviate from the spec, not only it hurts interoperability, and puts the burden on our users to know Node.js specific behavior. 
If we want that feature in the language, TC39 is the group in charge of that, I trust their judgment – and if that request has already been made and rejected for reasons we deemed unreasonable, that’s a different story, but I don’t think it’s the case. In the mean time, we can have this API available to loaders, so they can use it if they are not interested in following the spec.

Reply:
The loaders team recently allowed loaders to be dynamically registered at runtime, which is a decision I don't agree with. This allows anyone to register a loader at any time. Pandora's box of specification violations has already been opened, apparently with little fanfare. I think if we try to restrict this functionality of cached module eviction to "loaders only" what we will see is an npm module which registers itself as a loader and magically exports this capability to consumers.

Reply:
Perhaps that decision should be revisited, then - otherwise an onslaught of security vulnerabilities is likely to be forthcoming where random packages use the loader api, and this one, to do literally anything they want to an application.

Reply:
Is the intent here to support hot module reloading (HMR)? If that's the case, I think we should explore various alternatives first. For instance, adopting `vite`'s `import.meta.hot` approach could provide an alternative that does not violate the specification but carries the additional limitation that reloads cannot change the shape of the module namespace on updates. I'm a bit leary of landing any API here without carefully considering multiple approaches, and I absolutely agree with @aduh95 that, if at all possible, we should not deviate from the spec (at least by default).

(fwiw, I do see a path where a command line option could be used to enable non-standard HMR behavior)

Reply:
> It’s my opinion – and I believe the consensus in the project – that Node.js, as a JavaScript runtime, should parse JS source files according to the spec. Loaders can of course choose to parse them as something else, and deviate from the spec / follow a different one even.

I think the consensus has always been that Node follows spec _by default,_ but if users customize it behave otherwise then the consequences are the user’s responsibility. And the user accepts full responsibility for any dependencies they add; we don’t have any security sandbox protecting the “main” application from dependencies. Forcing this API to be accessed from the hooks thread just adds hassle for no benefit. If the user can do it at all, just let them do it.

Reply:
Yes, I am aware. I just think we need to step back and take a broader systematic view here. As it is, this *feels* a bit too ad hoc and fraught with potential issues.

Reply:
For instance, as an alternative, we could choose an approach where an explicitly opt-in command line flag is required to enable this kind of non-standard ESM behaviors.

```
node --esm-module-ext
```

This makes it absolutely clear that the individual *running* the node.js process is deciding on the non-standard behavior as opposed to some arbitrary library they are depending on.

Reply:
> Perhaps that decision should be revisited, then - otherwise an onslaught of security vulnerabilities is likely to be forthcoming where random packages use the loader api, and this one, to do literally anything they want to an application.

The built-in `node:*` module namespace is chocked full of security vectors and npm is already a huge hazard. Tinkering with module resolution is hardly a major concern given that `node:fs` and `node:child_process` exist. The real risk is ecosystem fragmentation, like what we have with CommonJS and ESM. This is comparable to `delete require.cache[key]` under CommonJS which wasn't even widely used in that ecosystem.

>HMR is already possible; that’s what Vite is doing

Vite's HMR implementation is a client (browser)-side feature. For SSR you have to use `await vite.ssrLoadModule('/main.js')` which operates outside of nodejs's module graph as I understand it. I might be wrong here because I personally don't use Vite.

[dynohot](https://github.com/braidnetworks/dynohot) implements `import.meta.hot` on the server-side as a nodejs loader. To my knowledge it's the only implementation that truly operates under nodejs's module graph. I implemented the es262 linking algorithm [to specification](https://github.com/braidnetworks/dynohot/blob/fb822d2022f9d71d9d7ab5377d5b5d55ddcb26a8/runtime/traverse.ts#L62-L67) and use some code transformations to tie it all together. I'd love to see such a feature in nodejs but without support from v8 you need a code transformation, otherwise there's no way to update the linking of an already resolved and imported name. It seems like code transpilation is something that is outside the scope of node's runtime (and thus we have the loader API).

Reply:
> Why should a user need to opt into this via flag when they’re already opting into it by using the API, either explicitly or via a dependency that they knowingly added?

Based on experience, I would wager very few users know what most of their dependencies are doing.

More importantly, I'm not saying no to this, I'm saying I'd like us to make sure we're we're being careful to not introduce new headaches we're might regret later. I guess what I'm missing is a higher level description of the use cases, how these various bits fit in with that, and a example that ties it all together. Has such a thing been written up?

Reply:
https://github.com/nodejs/node/issues/49442 has pieces of that but yes, a better writeup would be good. One of my notes on this PR is that it needs documentation, so that seems like a good place to start. @laverdet Could you please make the example written for the docs one that shows the most common use case?